### ISSUE TITLE
allow non-descriptor subtypes

### ISSUE THREAD

**Issue by choucavalier** [microsoft/pyright#11021]
### Description

consider the following example, where we want to define a `Protocol` for any object that has `system: str` and a `code: str` attributes:

```python
from typing import Protocol, runtime_checkable

from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass


class Base(MappedAsDataclass, DeclarativeBase):
    pass


@runtime_checkable
class ConceptProto(Protocol):
    @property
    def system(self) -> str: ...

    @property
    def code(self) -> str: ...


class Concept(Base):
    __tablename__ = "concept"

    system: Mapped[str]
    code: Mapped[str]


def fun(c: ConceptProto):
    print(c.system, c.code)


fun(Concept(system="SNOMED", code="123456789"))
```

this currently fails with pyright:

```
#########################/proto.py
  #########################/proto.py:30:5 - error: Argument of type "Concept" cannot be assigned to parameter "c" of type "ConceptProto" in function "fun"
    "Concept" is incompatible with protocol "ConceptProto"
      "system" is an incompatible type
        "Mapped[str]" is not assignable to "str"
      "code" is an incompatible type
        "Mapped[str]" is not assignable to "str" (reportArgumentType)
1 error, 0 warnings, 0 notes
```

it would be nice if sqlalchemy models could be used within that context, and manipulated by duck-typed libs without ever having to import sqlalchemy

you can see the initial discussion with the maintainer of sqlalchemy here: https://github.com/sqlalchemy/sqlalchemy/discussions/12889
